ddtrace/debugging/_rc.py

```diff
@@ -0,0 +1,220 @@
+            args["func_qname"] = attribs["where"].get("method") or attribs["where"]["methodName"]
+
+        return ProbeType(**args)
+
+    elif _type == "metricProbes":
+        return MetricProbe(
+            probe_id=_id,
+            active=attribs["active"],
+            tags=dict(_.split(":", maxsplit=1) for _ in attribs.get("tags", [])),
+            source_file=attribs["where"]["sourceFile"],
+            line=int(attribs["where"]["lines"][0]),
+            name=attribs["metricName"],
+            kind=attribs["kind"],
+        )
+
+    raise ValueError("Unknown probe type: %s" % _type)
+
+
+@attr.s
+class DebuggingRC(object):
+    service_name = attr.ib(type=str)
+    _config_endpoint = attr.ib(type=str, init=False)
+    _url = attr.ib(type=str, factory=get_trace_url)
+    _connect = attr.ib(type=Connector, init=False)
+    _req_payload = attr.ib(type=str, init=False)
+
+    def __attrs_post_init__(self):
+        # type: () -> None
+        self._config_endpoint = "/v0.7/config"
+        self._connect = connector(self._url, timeout=config.config_timeout)
+        self._req_payload = json.dumps(
+            {
+                "client": {
+                    "id": "beb0d6a3-75b0-424f-bbb9-98181545f731",
+                    "name": "dd-trace-py",
+                    "products": ["LIVE_DEBUGGING"],
+                    "version": __version__,
+                    "state": {},
+                    "is_tracer": True,
+                    "client_tracer": {
+                        "runtime_id": get_runtime_id(),
+                        "language": "python",
+                        "tracer_version": __version__,
+                        "service": self.service_name,
+                        "env": tracer_config.env,
+                        "app_version": tracer_config.version,
+                    },
+                },
+            }
+        )
+
+    @_filter_by_env_and_version
+    def get_probes(self):
+        # type: () -> Optional[Iterable[Probe]]
+        try:
+            with self._connect() as conn:
+                log.debug("Sending RCM request to the agent: %s", self._req_payload)
+                conn.request("POST", self._config_endpoint, self._req_payload)
+                resp = compat.get_connection_response(conn)
+                if resp.status == 200:
+                    try:
+                        max_size = config.max_payload_size
+                        raw_data = resp.read(max_size + 1)
+                        if len(raw_data) > max_size:
+                            log.error("Configuration payload size is too large (max: %d)", max_size)
+                            return None
+                        data = [
+                            json.loads(six.ensure_text(base64.b64decode(f["raw"])))
+                            for f in json.loads(six.ensure_text(raw_data))["target_files"]
+                        ]
+                    except KeyError:
```

Comment:
Would we possibly want a debug log here? Or do we think it'd just mislead people whenever they see it and there's nothing that would cause this to happen other than the caches being cold?

Reply:
I think the other branches are instrumented enough to allow us to conclude what is happening here